src/Aspire.Hosting/Dashboard/proto/resource_service.proto

```diff
@@ -0,0 +1,170 @@
+    //
+    // If the set of commands changes over time, use the "commands" property
+    // of the Resource itself.
+    repeated ResourceCommandRequest commands = 3;
+}
+
+////////////////////////////////////////////
+
+message EnvironmentVariable {
+    string name = 1;
+    optional string value = 2;
+    bool is_from_spec = 3;
+}
+
+message Endpoint {
+    string name = 1;
+    optional string allocated_address = 2;
+    optional int32 allocated_port = 3;
+    optional string http_address = 4;
+}
+
+message StringArray {
+    repeated string values = 1;
+}
+
+message AdditionalData {
+    // Name of the data item, e.g. "id", "image", "path", ...
+    string name = 1;
+    // Optional namespace, e.g. "container", "executable", "project", ...
+    optional string namespace = 2;
+    // Optional display name, may be localized
+    optional string display_name = 3;
+    // A single value will be most common. Also support lists, to avoid escaping.
+    oneof kind {
+        string value = 4;
+        StringArray values = 5;
+    }
+}
+
+message ResourceId {
+    string uid = 1;
+    // TODO do we need resource_type to make unique names? if not, inline ResourceId type as string.
+    string resource_type = 2;
+}
+
+// Models the full state of an resource (container, executable, project, etc) at a particular point in time.
+message ResourceSnapshot {
+    ResourceId resource_id = 1;
+    string display_name = 2;
+    optional string state = 3;
+    optional google.protobuf.Timestamp created_at = 4;
+    repeated EnvironmentVariable environment = 5;
+    optional int32 expected_endpoints_count = 6;
+    repeated Endpoint endpoints = 7;
+    repeated ResourceCommandRequest commands = 8;
+
+    // TODO potentially remove endpoints/env vars/anything that might not be general purpose
+
+    // List of additional data, as name/value pairs.
+    // For:
+    // - Containers: image, container_id, ports
+    // - Executables: process_id, executable_path, working_directory, arguments
+    // - Projects: process_id, project_path
+    repeated AdditionalData additional_data = 9;
+}
+
+////////////////////////////////////////////
+
+// Models a snapshot of resource state
+message WatchResourcesSnapshot {
+    repeated ResourceSnapshot resources = 1;
+    repeated ResourceType types = 2;
+}
+
+////////////////////////////////////////////
+
+message ResourceDeletion {
+    ResourceId resource_id = 1;
+}
+
+message WatchResourcesChange  {
+    oneof kind {
+        ResourceDeletion delete = 1;
+        ResourceSnapshot upsert = 2;
+    }
+}
+
+message WatchResourcesChanges {
+    repeated WatchResourcesChange value = 1;
+}
+
+////////////////////////////////////////////
+
+// Initiates a subscription for data about resources.
+message WatchResourcesRequest {
+    // True if the client is establishing this connection because a prior one closed unexpectedly.
+    optional bool is_reconnect = 1;
```

Comment:
How will the result of the request change if `is_recconect` is true? Does this imply the implementer of `DashboardService` is stateful?

Reply:
The idea is just to allow the server to know if the client is experiencing disconnections. This would go in a log or telemetry or something like that.